File: examples/posix-thread/example04.cpp
// =================================================================
//
// File: example04.cpp
// Author: Pedro Perez
// Description: This file contains the code that searches for the
// 				smallest value stored in an array using pthreads.
//              To compile: g++ example4.cpp -lpthread
//
// Copyright (c) 2022 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include <iostream>
#include <iomanip>
#include <climits>
#include <unistd.h>
#include <pthread.h>
#include "utils.h"

using namespace std;

const int SIZE = 100000000; //1e8
const int THREADS = 4;

typedef struct {
  int start, end; // [start, end)
  int *arr;
} Block;

void* minValue(void* param) {
  int *acum;
  Block *block;
  int i;

  block = (Block *) param;
  acum = new int;
  (*acum) = INT_MAX;
  for (i = block->start; i < block->end; i++) {
    (*acum) += min((*acum), block->arr[i]);
  }
  return ( (void**) acum );
}

int main(int argc, char* argv[]) {
	int *a, block_size, i, j, pos;
	double ms, result, *acum;
  Block blocks[THREADS];
  pthread_t tids[THREADS];

	a = new int[SIZE];
	fill_array(a, SIZE);
	display_array("a", a);

  srand(time(0));
	pos = rand() % SIZE;
	printf("Setting value 0 at %i\n", pos);
	a[pos] = 0;

  block_size = SIZE / THREADS;
  for (i = 0; i < THREADS; i++) {
      blocks[i].arr = a;
      blocks[i].start = i * block_size;
      if (i != (THREADS - 1)) {
          blocks[i].end = (i + 1) * block_size;
      } else {
          blocks[i].end = SIZE;
      }
  }

	std::cout << "Starting...\n";
	ms = 0;
    for (j = 0; j < N; j++) {
        start_timer();

        result = 0;
        for (i = 0; i < THREADS; i++) {
            pthread_create(&tids[i], NULL, minValue, (void*) &blocks[i]);
        }
        for (i = 0; i < THREADS; i++) {
            pthread_join(tids[i], (void**) &acum);
            result += (*acum);
            free(acum);
        }

        ms += stop_timer();
    }
	std::cout << "min = " << setprecision(5) << result << "\n";
	std::cout << "avg time =  " << setprecision(5) << (ms / N) << "\n";

	delete [] a;
	return 0;
}

File: examples/processes-python/example04.py
## =================================================================
##
## File: example04.py
## Author: Pedro Perez
## Description: This file contains the code that searches for the
## 				smallest value stored in an array using processes
##              in Python.
##
## Copyright (c) 2022 by Tecnologico de Monterrey.
## All Rights Reserved. May be reproduced for any non-commercial
## purpose.
##
## =================================================================

import utils
import time
import sys
from random import seed
from random import random
import multiprocessing as mp
import threading
from random import randrange

# CORES = mp.cpu_count() 
CORES = 4
SIZE = 100000000 ##1e8

def partialMinArray(start, end, array, queue, ):
    result = sys.maxsize
    for i in range(start, end):
        result = min(result, array[i])
    queue.put(result)

if __name__ == "__main__":
    array = [0] * SIZE

    utils.randomArray(array)
    utils.displayArray("array", array)

    pos = randrange(SIZE)
    array[pos] = 0

    blockSize = SIZE // CORES

    print("Starting...")
    startTime = endTime = ms = 0
    for j in range(utils.N):
        startTime = time.time() * 1000

        queue = mp.SimpleQueue()
        processes = list()
        for i in range(CORES):
            start = i * blockSize
            if i != (CORES - 1):
                end = (i + 1) * blockSize
            else:
                end = SIZE
            process = mp.Process(target=partialMinArray, args=(start, end, array, queue,))            
            processes.append(process)
            process.start()
                
        result = 0
        for i in range(CORES):
            result = min(result, queue.get())

        endTime = time.time() * 1000

        ms = ms + (endTime - startTime)

    print("minimum = ", result)
    print("avg time = ", (ms / utils.N), " ms")